### ISSUE TITLE
Illegal instruction on Haswell

### ISSUE THREAD

**Issue by yln** [StanfordPL/stoke#869]
I was able to successfully follow the `examples/tutorial` as described in the main README.
However, when I try to optimize `HighwayTreeHash` from [here](https://github.com/google/highwayhash/blob/master/highway_tree_hash.cc) the search fails with an "Illegal instruction" error.

I ran `make clean && make debug && make test` a couple of times. Two tests sporadically fail with "Illegal instruction":
- IntegrationTest.SearchInitZero
- IntegrationTest.SearchNonGoal

Attached are my config files and other (hopefully useful) debugging output:
[configure.sh](https://github.com/StanfordPL/stoke/files/235627/configure.sh.txt)
[cpuinfo](https://github.com/StanfordPL/stoke/files/235628/cpuinfo.txt)

[make_test](https://github.com/StanfordPL/stoke/files/235674/make_test.txt)
[make_test2](https://github.com/StanfordPL/stoke/files/235675/make_test2.txt)

[extract.conf](https://github.com/StanfordPL/stoke/files/235688/extract.conf.txt)
[testcase.conf](https://github.com/StanfordPL/stoke/files/235666/testcase.conf.txt)
[search.conf](https://github.com/StanfordPL/stoke/files/235667/search.conf.txt)

[HighwayTreeHash.s](https://github.com/StanfordPL/stoke/files/235684/HighwayTreeHash.s.txt)
[HighwayTreeHash.tc](https://github.com/StanfordPL/stoke/files/235690/HighwayTreeHash.tc.txt)

[stoke_search](https://github.com/StanfordPL/stoke/files/235672/stoke_search.txt)
[stoke_search2](https://github.com/StanfordPL/stoke/files/235673/stoke_search2.txt)

**Comment by bchurchill** [comment]
Thank you for reporting.  I'll try to investigate this soon.

**Comment by bchurchill** [comment]
What proportion of the time were you able to reproduce this bug (the HighwayTreeHash one)?  How long did you typically need to run the search to hit it?

**Comment by yln** [comment]
The search fails on every attempt (never successfully completes, never times out).
The [stoke_search*] files give examples of the searches. It usually runs between 0-3 seconds before the "Illegal instruction" error.

Let me know if/how I can provide more information.

**Comment by bchurchill** [comment]
Could you confirm which commit of stoke, x64asm and cpputil that triggered this bug?  (x64asm and cpputil are in the src/ext folder).

**Comment by bchurchill** [comment]
I can't seem to reproduce it, but I suspect that a probable cause are the CpuFlags that you have on your processor that we don't have (these are "hle", "rtm", "x2apic").  It's possible that stoke/x64asm thinks it can propose some instructions we've never properly tested in our sandbox.  This would also explain the root cause of #840.

### Things you can do:

1. (a) Try adding this to your config file:

`--cpu_flags { }`

Hopefully this will solve the problem, but it will also reduce the instructions you can use to a smaller base set.  

 (b) You can then add flags for useful instruction to this set; a good common choice might be:

`--cpu_flags { sse sse2 ssse3 sse4_1 sse4_2 avx avx2 popcnt cmov bmi1 bmi2 }`

hopefully this will work too.  

2. If you're able to find a minimal choice for --cpu_flags that causes this error to happen, please let us know and maybe we can fix the root cause.  My guess is that either of `--cpu_flags { hle }`, `--cpu_flags { rtm }` will cause problems.  It would be helpful for us if you can build in debug mode and tell us the malformed instruction, as has been done in #840.

**Comment by stefanheule** [comment]
To add to this, it would be helpful if you could run the tests (`make test`) and tell us if they are passing or not.  The latest version of stoke should pass all tests, and we do have a fuzzer for the sandbox.  But as @bchurchill pointed out, we have slightly less fancy hardware, so we can't test the sandbox on some instructions, so you might get a test failure that would immediately tell us what's wrong.

**Comment by bchurchill** [comment]
I'm almost sure now the problem are the HLE/RTM instructions, which x64asm thinks it support but won't actually work on linux.  This is consistent with #840; the disassembly shown there suggests that STOKE is trying to assemble "xacquire" and "xrelease", which I can't possibly imagine working without some kind of setup.

**Comment by bchurchill** [comment]
It's possible we just fixed this in #873.  Would you be willing to confirm?  Do a "git pull", make clean && make, and try it all again.

**Comment by yln** [comment]
`git pull && make clean && make debug && make test` reports the same two failing tests.

`--cpu_flags "{ hle }"` fails with the illegal instruction error.
`--cpu_flags "{ rtm }"` seems to work.

How do I tell stoke to persist the transformed program version so that I can run it in gdb/disassemble?
Or better question: what do I need to do to get debug output like in #840?

**Comment by yln** [comment]
_bump_

**Comment by bchurchill** [comment]
I would recommend, as a work around, just passing in something like 

`--cpu_flags { sse sse2 ssse3 sse4_1 sse4_2 avx avx2 popcnt cmov bmi1 bmi2 }`

or just do `--cpu_flags { }`.  It might be a while before we have a proper fix because we'll need to test on a processor with the hle flag.

To get the debug output as in #840, I believe you should be able to do exactly what he did: run the program in gdb, and, when it hits an IllegalInstruction, you'll see something like:

`0x00007ffff155402e in ?? ()`

Then, to disassemble, you copy this constant and run:

`disassemble /r 0x00007ffff155402e,+32`

I suspect the output you see will be very similar to his.

**Comment by yln** [comment]
Posting my output, maybe it helps.

```
Program received signal SIGILL, Illegal instruction.
0x00007ffff1b610ef in ?? ()
(gdb) disassemble /r 0x00007ffff1b610ef,+32
Dump of assembler code from 0x7ffff1b610ef to 0x7ffff1b6110f:
=> 0x00007ffff1b610ef:  f3 c4 42 99 a9 d3   repz vfmadd213sd %xmm11,%xmm12,%xmm10
   0x00007ffff1b610f5:  66 0f b6 e8 movzbw %al,%bp
   0x00007ffff1b610f9:  90  nop
   0x00007ffff1b610fa:  48 a3 70 c2 fe ff ff 7f 00 00   movabs %rax,0x7ffffffec270
   0x00007ffff1b61104:  48 89 e0    mov    %rsp,%rax
   0x00007ffff1b61107:  48 a3 30 c3 fe ff ff 7f 00 00   movabs %rax,0x7ffffffec330
End of assembler dump.
```

Similarly faulting instructions:

```
=> 0x00007ffff1b61031:  f2 c4 42 7d 59 dc   repnz vpbroadcastq %xmm12,%ymm11
=> 0x00007ffff1b612c8:  f7 fb   idiv   %ebx
=> 0x00007ffff1b61105:  f3 c4 e2 50 f3 d8   repz blsi %eax,%ebp
=> 0x00007ffff1b61133:  48 f7 f2    div    %rdx
=> 0x00007ffff1b61035:  f3 c5 fc 77 repz vzeroall
```

I am currently using the workaround, but can't seem to find any rewrites that result in faster execution.
Is this a property of my program input [HighwayTreeHash](https://github.com/google/highwayhash/blob/master/highway_tree_hash.cc) or am I doing something stupid?
My config: [search.conf](https://github.com/StanfordPL/stoke/files/253032/search.conf.txt)

**Comment by bchurchill** [comment]
I took a look at the config and it looks sane but there are a few things that can be improved that could make a big difference:
1. Set the total number of search iterations much higher.  For something of this size, 10^6 is an absolute minimum.  10^7 and 10^8 would be reasonable, but it's hard to say what a good stopping point is.
2. Use "--cost correctness+measured".  The "measured" cost function is much better for loops.
3. Since your code is mostly using AVX instructions, you can narrow the search space better with "--cpu_flags { avx avx2 bmi bmi2 }".  Many of the other flags are just going to slow the search down.
4. Run STOKE many times.  A single run of STOKE samples from a distribution of possible outputs.  Some are good, some are bad.  For example, you might want to do 100 runs with 10^6 or 10^7 iterations each.  Then you'll need to benchmark the output and pick the best ones.  It's best to have a benchmarking script to automate this.
5. Be aware that the current "verification" that's being done is only on testcases; it's not real verification.  To get real verification, I would recommend "--strategy bounded" with "--bound 1" to get started.  You might need some instructions are unsupported; if so, let us know and we can work together to add them.  "--validator_must_support" can be used to limit the search space to supported instructions by the validator, but it won't help if your target has unsupported instructions.

I know this becomes a somewhat long exercise, and it's not guaranteed to work.  I can't say I've looked at your particular program, so I don't have any intuition for what optimizations might exist or if STOKE will find them.  The above procedure has worked for us on other examples comparable to the size of your program, so maybe it will work for you too.

If you try these things out, and you're still stuck, we can look at it further.  If that's needed, it would be helpful if you could provide a benchmarking script (e.g. the one from step 3) that will run a rewrite and tell us how well it performs (you might use "stoke replace", for example).

**Comment by bchurchill** [comment]
I've played with this example a little more, and it looks like STOKE certainly _thinks_ it can reduce the cost of the program significantly, even with just 10000 iterations.  You'll need to benchmark it to be sure.  However, it may not be giving correct rewrites.  Verification looks like it will be hard because of unsupported instructions in the target, so you may need to hold off on that for now.

**Comment by yln** [comment]
Thank you for all your help. I apologize for the tardy reply.

Just to make sure, with 1. you mean changing

```
# --cycle_timeout <string>
# Default: "10000, 10000, 10000, 10000, 10000, 2**i * 1000"
```

 to

```
--cycle_timeout "10**7, 10**7, 10**7, 10**7, 10**7, 2**i * 10**6"
```

Correct?

Will try to automate the process so I can benchmark different versions and pick the fastest.

**Comment by yln** [comment]
Weirdly, I now consistently get errors when trying to patch up my original binary with the optimized assembly.

One example:

```
$ stoke replace --config replace.conf 

Errors:

  --rewrite <path/to/file.s>
      Error reading value: Failed to parse line 88: vcvtss2sil es:0x2(%rax,%rax,1), %ecx                #  80    0x400ea5  7      OPC=vcvtss2sil_r32_m32            
      Message: Could not parse operand 1: Don't know how to parse this!
      Failed to parse line 89: vblendps $0xff, es:0xe(%rax,%rax,1), %xmm10, %xmm5  #  81    0x400eac  9      OPC=vblendps_xmm_xmm_m128_imm8    
      Message: Could not parse operand 2: Don't know how to parse this!
      Failed to parse line 98: xaddl %esi, es:(%rax,%rax,1)                        #  90    0x400ef0  5      OPC=xaddl_m32_r32                 
      Message: Could not parse operand 2: Don't know how to parse this!
      Failed to parse line 100: vpaddd es:0x1(%rax,%rax,1), %xmm9, %xmm11           #  92    0x400efc  7      OPC=vpaddd_xmm_xmm_m128           
      Message: Could not parse operand 1: Don't know how to parse this!
      Failed to parse line 107: salw $0x5, es:0x5(%rax,%rax,1)                      #  99    0x400f22  7      OPC=salw_m16_imm8                 
      Message: Could not parse operand 2: Don't know how to parse this!
      Failed to parse line 115: vmovq es:0x2(%rax,%rax,1), %xmm5                    #  107   0x400f4f  8      OPC=vmovq_xmm_m64                 
      Message: Could not parse operand 1: Don't know how to parse this!
      Failed to parse line 117: vpmovsxwd es:0x6(%rax,%rax,1), %xmm4                #  109   0x400f5c  8      OPC=vpmovsxwd_xmm_m64             
      Message: Could not parse operand 1: Don't know how to parse this!
```

I am working off 6c064918 (just did a `git pull`) and gcc version `gcc (Ubuntu 4.9.3-8ubuntu2~14.04) 4.9.3`.

I attached the original and the optimized assembly file.
[HighwayTreeHash.s.orig](https://github.com/StanfordPL/stoke/files/271481/HighwayTreeHash.s.txt)
[HighwayTreeHash.s.opt](https://github.com/StanfordPL/stoke/files/271485/HighwayTreeHash.s.txt)

**Comment by stefanheule** [comment]
Yes.  The idea of the cycle timeout is to restart the search if it gets stuck, but this is more relevant for synthesis than optimization.  You may want to just use `--cycle_timeout N --timeout_iterations N` for some large N (the number of iterations you want to do).  I will look into making these defaults better.

I'm will also look into the parsing errors you are getting.

**Comment by stefanheule** [comment]
@yln, I tracked down the problem you encountered with the parsing errors and fixed it in #908.  You will have to run STOKE again to find a new rewrite though, as the problem was with the STOKE search procedure.

**Comment by bchurchill** [comment]
You might also be interested in the feature Stefan recently implemented
which saves many results into a folder, instead of just outputting one.  It
doesn't remove the need to run STOKE more than once, but might make
scripting easier.

**Comment by stefanheule** [comment]
In case it isn't clear already, @bchurchill is talking about the feature implemented in #900.  It will save many rewrites to a folder from a single search, which you could then take and benchmark each of them to see which one is fastest.

**Comment by bchurchill** [comment]
If anything else is needed for this ticket, please let us know.
